### PR TITLE
Design for Same-Table-JOINs

### PR DESCRIPTION
**Pull request by acarbonetto**
### Description
Added documentation for a same-table join proposal. 

related: https://opensearch.org/docs/latest/field-types/join/
 
### Issues Resolved
* https://github.com/opensearch-project/sql/issues/1441
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: docs/dev/opensearch-same-table-join-query.md [hunk 1]
```diff
@@ -0,0 +1,1252 @@
+# Overview
+
+The OpenSearch SQL plugin provides a thin-client language-processing engine from SQL to an OpenSeach instance.  The SQL
+language is a relational language that provides the JOIN feature to relate tables together by field or expression.  OpenSearch
+offers a similar relation mapping within an index, that allows users to relate same-index documents using a parent-child
```
Comment (forestmvey) [inline docs/dev/opensearch-same-table-join-query.md:5]: Maybe we can link to the documentatin page for users looking for a more in-depth description of the mapping type.

File: docs/dev/opensearch-same-table-join-query.md [hunk 2]
```diff
@@ -0,0 +1,1252 @@
...
+  * [Support for _routing field in SQL queries](https://github.com/opendistro-for-elasticsearch/sql/issues/1151)
+* [Nested Design](https://github.com/Bit-Quill/opensearch-project-sql/pull/262)
+* [opensearch.org: Join field type](https://opensearch.org/docs/latest/field-types/join/)
+* [PariQL](https://partiql.org/)
+
+## 3. Assumptions/Dependencies
+
+* _shard must be defined for any query using a same-table join
+* Same-Table joins requires that the following work is complete:
+    * Table/Indices can be aliased, for example: `SELECT t.fieldA, t.fieldB FROM table t WHERE t.fieldA = t.fieldB`
```
Comment (forestmvey) [inline docs/dev/opensearch-same-table-join-query.md:37]: Is there an issue tracking this?

File: docs/dev/opensearch-same-table-join-query.md [hunk 3]
```diff
@@ -0,0 +1,1252 @@
...
+### 3.3. Storage Type
+
+* `reference to opensearch doc`
+* `Describe how to define the JOIN-relation`
+* `What the datatype can do`
+* `What the datatype cannot do`
+
+## 4. Use Cases
+
+Use Cases have examples that is defined in the Annex.
```
Comment (forestmvey) [inline docs/dev/opensearch-same-table-join-query.md:64]: Maybe separate use cases and examples?

Comment (acarbonetto) [reply docs/dev/opensearch-same-table-join-query.md:64]: Sorry.  This should say use cases with examples.

File: docs/dev/opensearch-same-table-join-query.md (hunk 1 above)
Comment (acarbonetto) [reply docs/dev/opensearch-same-table-join-query.md:5]: Updated.  Thanks

File: docs/dev/opensearch-same-table-join-query.md [hunk 4]
```diff
@@ -0,0 +1,1368 @@
...
+* Performance on the OpenSearch side
+* Multiple-shard queries using `join`
+* `join` using lateral-table support [PartiQL support in OpenSearch SQL V2](https://github.com/opensearch-project/sql/issues/1104)
+* Pagination using `join` or other complex types
+
+### 3.2 Design Decisions
+
+* JOIN-relations architecture is built to work in combination with the `nested` and `object` data types.  This allows enables
+  both same-table joins and nested object (using PartiQL) to share execution paths and syntax.  This also enables a shared
+  solution to expand support for `lateral` tables and enable `lateral join` and `lateral nested` queries.
```
Comment (forestmvey) [inline docs/dev/opensearch-same-table-join-query.md:55]:
```suggestion
  solution can expand support for `lateral` tables and enable `lateral join` and `lateral nested` queries.
```

File: docs/dev/opensearch-same-table-join-query.md [hunk 5]
```diff
@@ -0,0 +1,1368 @@
...
+                }
+            }
+        ]
+    }
+}
+```
+
+### 4.3 Case SELECT... FROM... JOIN ON... WHERE parent relation
+
+Summary:
```
Comment (forestmvey) [inline docs/dev/opensearch-same-table-join-query.md:292]: Filling in at least a small overview for each of these would examples would be useful.

File: docs/dev/opensearch-same-table-join-query.md [hunk 6]
```diff
@@ -0,0 +1,1368 @@
...
+
+#### 6.1.4. Same-table JOIN using `parent_id` relation:
+```sql
+FROM table AS tableAlias
+JOIN table AS joinTableAlias ON (joinTableAlias.relation.parent = 'myId')
+```
+Where `relation` is the data field mapped as a `join` type, and we want to join on the parent documents with `_id` equal
+to the string `myId`.
+
+### 6.1 Presentation
```
Comment (forestmvey) [inline docs/dev/opensearch-same-table-join-query.md:976]: Numbering got a bit screwed up here for the sections. Maybe a section named `Low-level Design` would be good to have these diagrams in?

File: docs/dev/opensearch-same-table-join-query.md [hunk 7]
```diff
@@ -0,0 +1,1368 @@
...
+
+The following diagram explains the proposed sequence to create JOIN request to OpenSearch
+
+### 6.1 Logical Plan Changes - Presentation
```
Comment (forestmvey) [inline docs/dev/opensearch-same-table-join-query.md:980]: Maybe we can change the `push down OP` state change to `execute query` or something like this. Push down happens during logical plan optimization:
[LogicalPlanOptimizer.java#L67-L73](https://github.com/opensearch-project/sql/blob/main/core/src/main/java/org/opensearch/sql/planner/optimizer/LogicalPlanOptimizer.java#L67-L73) 
Starting from the deepest part of the `Logical Plan Tree` each `LogicalPlan` is optimized/pushed down. After push down most LogicalPlans are removed from the tree, except generally `LogicalProject`, a logical plan to instantiate an operator for post-processing, and the `LogicalRelation` will be replaced with the `OpenSearchIndexScanBuilder`. So we could remove `LogicalFilter` and `LogicalSort` from the middle composite state.

Comment (acarbonetto) [reply docs/dev/opensearch-same-table-join-query.md:980]: I see what you're saying.  The "push down" terminology is a little ambiguous.  I think what we're really talking about are the transitions between states.  I've changes this to `Build Query` which I hope makes sense?

File: docs/dev/opensearch-same-table-join-query.md (hunk 6 above)
Comment (acarbonetto) [reply docs/dev/opensearch-same-table-join-query.md:976]: Thanks.  Fixed.

File: docs/dev/opensearch-same-table-join-query.md [hunk 8]
```diff
@@ -0,0 +1,1368 @@
...
+| Allow for child-parent joins                                           | must have    | P0    |
+| Retrieve inner_hits                                                    | must have    | P0    |
+| Allow for parent_id joins                                              | nice to have | P1    |
+| Aggregation on children                                                | nice to have | P2    |
+| Sorting on children                                                    | nice to have | P2    |
+| Documentation for same-table Joins                                     | must have    | P0    |
+
+_Phase `*`_ denotes dependencies that must be completed prior to starting this work.
+
+### 5.1 Release Schedule
```
[the remark this answers is not in the document]
Comment (acarbonetto) [reply docs/dev/opensearch-same-table-join-query.md:883]: I'm going to keep this here... it's really to talk about what's a `must have` for the release.

**Pull request merged**